### ISSUE TITLE
ed.evaluate keep adding to graph

### ISSUE THREAD

**Issue by jingsliu** [blei-lab/edward#576]
Hi, I noticed that running the evaluate method keeps adding tensors to the tensorflow graph. Seems to be caused by line 122

"""
 y_pred = tf.add_n(y_pred) / tf.cast(n_samples, tf.float32)
"""

It's probably ok if the evaluation is only ran once per model; but if I need to monitor validation error while training, the memory blows up soon especially when validation set is large. Is it possible to remove the added tensors, or simply use python native functions instead of tf functions for this part?

Thanks!

**Comment by dustinvtran** [comment]
Great question. `ed.evaluate` and `ed.ppc` do two things: they build the graph of necessary computation and then they run it to output NumPy arrays. 

One difficulty is that the methods run the graph during an immediate step in order to obtain posterior predictive samples. This is done for efficiency reasons, because otherwise instead of fetching a random variable to get 100 samples, we'd have to copy it 100 times (this is impractical for large models). 

If we want to separate the graph-building, there are two options as you mention:

1. After fetching the samples, we force everything to use native Python functions. This includes the line you mention as well as the metrics. This seems rather unfortunate. 
2. Alternatively—and which is perhaps more TensorFlow-friendly—is to cache this latter part of the TensorFlow graph and fetch it. For caching, we can check if a tensor with a given metric name exists (e.g., with `graph.get_tensor_by_name()`). If it does, fetch it while feeding a placeholder for `y_pred`; otherwise, build it given the placeholder and then fetch it.

I recommend 2. Contributions welcome!
